blobit-core/src/main/java/org/blobit/core/cluster/HerdDBMetadataStorageManager.java

```diff
@@ -442,55 +442,87 @@ private Connection getConnectionForBucketTableSpace(BucketMetadata bucket) throw
+                        /*
+                         * We should fail on concurrent creation or creation failed and tablespace still
+                         * doesn't exists, throw original error
+                         */
+                        throw e;
+                    }
                 }
             }
         }
     }
 
+    private boolean checkTablespaceExistence(String schema, DatabaseMetaData metaData) throws SQLException {
+        try (ResultSet schemas = metaData.getSchemas(null, schema);) {
+            return schemas.next();
+        }
+    }
+
     private boolean existsTablespaceForBucket(String schema) throws SQLException {
         if (!useTablespaces) {
             return true;
         }
         try (Connection connection = datasource.getConnection()) {
             connection.setSchema(bucketsTablespace);
-            DatabaseMetaData metaData = connection.getMetaData();
-            try (ResultSet schemas = metaData.getSchemas(null, schema);) {
-                return schemas.next();
-            }
+            final DatabaseMetaData metaData = connection.getMetaData();
+            return checkTablespaceExistence(schema, metaData);
         }
     }
 
     private void ensureTable(String schema, String name, String createSql) throws SQLException {
+        ensureTable(schema, name, createSql, true);
+    }
+
+    private void ensureTable(String schema, String name, String createSql, boolean failOnConcurrentCreation)
+            throws SQLException {
         try (Connection connection = datasource.getConnection()) {
             if (useTablespaces) {
                 connection.setSchema(schema);
             }
-            DatabaseMetaData metaData = connection.getMetaData();
-            boolean existTable;
-            try (ResultSet rs = metaData.getTables(null, null, name, null)) {
-                existTable = rs.next();
-            }
+            final DatabaseMetaData metaData = connection.getMetaData();
+            boolean existTable = checkTableExistence(name, metaData);
             if (!existTable) {
                 try (Statement s = connection.createStatement();) {
                     s.executeUpdate(createSql);
+                } catch (SQLException e) {
+                    /* Check if a concurrent process already created the table */
+                    if (failOnConcurrentCreation || !checkTableExistence(name, metaData)) {
+                        /*
+                         * We should fail on concurrent creation or creation failed and table still
+                         * doesn't exists, throw original error
+                         */
+                        throw e;
+                    }
                 }
             }
         }
     }
 
+    private boolean checkTableExistence(String table, DatabaseMetaData metaData) throws SQLException {
+        try (ResultSet rs = metaData.getTables(null, null, table, null)) {
```

Comment:
I am not sure about this check.
You can have the same table name in different tablespaces.

I guess this won't work

Reply:
Actually is just a move around of existing code just to avoid code duplication